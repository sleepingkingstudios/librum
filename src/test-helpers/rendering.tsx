import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
