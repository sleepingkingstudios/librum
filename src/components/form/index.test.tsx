import * as React from 'react';
import { Field } from 'formik';
import { faRadiation } from '@fortawesome/free-solid-svg-icons';

import '@testing-library/jest-dom';
import { waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Form } from './index';
import type {
  Mutation,
  MutationStatus,
  QueryParams,
  UseMutation,
} from './types';
import type { FetchPromise } from '@store/types';
import { render } from '@test-helpers/rendering';
import type { Middleware } from '@utils/middleware';

describe('<Form />', () => {
  it('should submit the form', async () => {
    const mutation: Mutation = jest.fn();
    const mutationStatus: MutationStatus = { isLoading: false };
    const useMutation: UseMutation = () => [mutation, mutationStatus];

    const { getByRole } = render(
      <Form initialValues={{}} useMutation={useMutation}>
        <button type="submit">Submit</button>
      </Form>
    );

    userEvent.click(getByRole('button', { name: 'Submit'}));

    await waitFor(() => {
      expect(mutation).toHaveBeenCalledWith({});
    });
  });

  it('should match the snapshot', () => {
    const mutation: Mutation = jest.fn();
    const mutationStatus: MutationStatus = { isLoading: false };
    const useMutation: UseMutation = () => [mutation, mutationStatus];

    const { asFragment } = render(
      <Form initialValues={{}} useMutation={useMutation}>
        <button type="submit">Submit</button>
      </Form>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('when the form is loading', () => {
    it('should match the snapshot', () => {
      const mutation: Mutation = jest.fn();
      const mutationStatus: MutationStatus = { isLoading: true };
      const useMutation: UseMutation = () => [mutation, mutationStatus];

      const { asFragment } = render(
        <Form initialValues={{}} useMutation={useMutation}>
          <button type="submit">Submit</button>
        </Form>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with className: value', () => {
    it('should match the snapshot', () => {
      const mutation: Mutation = jest.fn();
      const mutationStatus: MutationStatus = { isLoading: false };
      const useMutation: UseMutation = () => [mutation, mutationStatus];

      const { asFragment } = render(
        <Form
          className="custom-form"
          initialValues={{}}
          useMutation={useMutation}
        >
          <button type="submit">Submit</button>
        </Form>
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with loadingAnimation: value', () => {
    describe('when the form is loading', () => {
      it('should match the snapshot', () => {
        const mutation: Mutation = jest.fn();
        const mutationStatus: MutationStatus = { isLoading: true };
        const useMutation: UseMutation = () => [mutation, mutationStatus];

        const { asFragment } = render(
          <Form
            loadingAnimation="pulse"
            initialValues={{}}
            useMutation={useMutation}
          >
            <button type="submit">Submit</button>
          </Form>
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with loadingIcon: value', () => {
    describe('when the form is loading', () => {
      it('should match the snapshot', () => {
        const mutation: Mutation = jest.fn();
        const mutationStatus: MutationStatus = { isLoading: true };
        const useMutation: UseMutation = () => [mutation, mutationStatus];

        const { asFragment } = render(
          <Form
            loadingIcon={faRadiation}
            initialValues={{}}
            useMutation={useMutation}
          >
            <button type="submit">Submit</button>
          </Form>
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with loadingMessage: value', () => {
    describe('when the form is loading', () => {
      it('should match the snapshot', () => {
        const mutation: Mutation = jest.fn();
        const mutationStatus: MutationStatus = { isLoading: true };
        const useMutation: UseMutation = () => [mutation, mutationStatus];

        const { asFragment } = render(
          <Form
            loadingMessage="Activating Reactor..."
            initialValues={{}}
            useMutation={useMutation}
          >
            <button type="submit">Submit</button>
          </Form>
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with middleware', () => {
    const middleware: Middleware<
      Record<string, unknown>,
      FetchPromise<Record<string, unknown>>
    > = (nextFn, data) => nextFn({ ...data, secret: '12345' });

    it('should submit the form', async () => {
      const mutation: Mutation = jest.fn();
      const mutationStatus: MutationStatus = { isLoading: false };
      const useMutation: UseMutation = () => [mutation, mutationStatus];

      const { getByRole } = render(
        <Form initialValues={{}} middleware={middleware} useMutation={useMutation}>
          <button type="submit">Submit</button>
        </Form>
      );

      userEvent.click(getByRole('button', { name: 'Submit'}));

      await waitFor(() => {
        expect(mutation).toHaveBeenCalledWith({ secret: '12345' });
      });
    });
  });

  describe('with params', () => {
    const params: QueryParams = { id: 0 };

    it('should submit the form', async () => {
      const mutation: Mutation = jest.fn();
      const mutationStatus: MutationStatus = { isLoading: false };
      const useMutation: UseMutation = () => [mutation, mutationStatus];

      const { getByRole } = render(
        <Form params={params} initialValues={{}} useMutation={useMutation}>
          <button type="submit">Submit</button>
        </Form>
      );

      userEvent.click(getByRole('button', { name: 'Submit'}));

      await waitFor(() => {
        expect(mutation).toHaveBeenCalledWith(params);
      });
    });
  });

  describe('with form inputs', () => {
    const Fields = (): JSX.Element => (
      <React.Fragment>
        <label>
          Launch Site
          <Field id="launchSite" name="launchSite" />
        </label>

        <label>
          Mission Name
          <Field id="mission.name" name="mission.name" />
        </label>
      </React.Fragment>
    );
    const defaultValues = {
      launchSite: '',
      mission: {
        name: '',
      }
    };

    it('should submit the form', async () => {
      const mutation: Mutation = jest.fn();
      const mutationStatus: MutationStatus = { isLoading: false };
      const useMutation: UseMutation = () => [mutation, mutationStatus];

      const { getByRole } = render(
        <Form initialValues={defaultValues} useMutation={useMutation}>
          <Fields />

          <button type="submit">Submit</button>
        </Form>
      );

      userEvent.click(getByRole('button', { name: 'Submit'}));

      await waitFor(() => {
        expect(mutation).toHaveBeenCalledWith(defaultValues);
      });
    });

    describe('with middleware', () => {
      const middleware: Middleware<
        Record<string, unknown>,
        FetchPromise<Record<string, unknown>>
      > = (nextFn, data) => nextFn({ ...data, secret: '12345' });

      it('should submit the form', async () => {
        const mutation: Mutation = jest.fn();
        const mutationStatus: MutationStatus = { isLoading: false };
        const useMutation: UseMutation = () => [mutation, mutationStatus];

        const { getByRole } = render(
          <Form initialValues={defaultValues} middleware={middleware} useMutation={useMutation}>
            <button type="submit">Submit</button>
          </Form>
        );

        userEvent.click(getByRole('button', { name: 'Submit'}));

        await waitFor(() => {
          expect(mutation).toHaveBeenCalledWith({ ...defaultValues, secret: '12345' });
        });
      });
    });

    describe('when the user inputs values', () => {
      it('should submit the form', async () => {
        const mutation: Mutation = jest.fn();
        const mutationStatus: MutationStatus = { isLoading: false };
        const useMutation: UseMutation = () => [mutation, mutationStatus];
        const expectedValues = {
          launchSite: 'KSC',
          mission: {
            name: 'Warlock IV',
          },
        };

        const { getByRole, getByLabelText } = render(
          <Form initialValues={defaultValues} useMutation={useMutation}>
            <Fields />

            <button type="submit">Submit</button>
          </Form>
        );

        userEvent.type(getByLabelText('Launch Site'), 'KSC');

        userEvent.type(getByLabelText('Mission Name'), 'Warlock IV');

        userEvent.click(getByRole('button', { name: 'Submit'}));

        await waitFor(() => {
          expect(mutation).toHaveBeenCalledWith(expectedValues);
        });
      });

      describe('with params', () => {
        const params: QueryParams = { id: 0 };

        it('should submit the form', async () => {
          const mutation: Mutation = jest.fn();
          const mutationStatus: MutationStatus = { isLoading: false };
          const useMutation: UseMutation = () => [mutation, mutationStatus];
          const expectedValues = {
            launchSite: 'KSC',
            mission: {
              name: 'Warlock IV',
            },
            ...params,
          };

          const { getByRole, getByLabelText } = render(
            <Form params={params} initialValues={defaultValues} useMutation={useMutation}>
              <Fields />

              <button type="submit">Submit</button>
            </Form>
          );

          userEvent.type(getByLabelText('Launch Site'), 'KSC');

          userEvent.type(getByLabelText('Mission Name'), 'Warlock IV');

          userEvent.click(getByRole('button', { name: 'Submit'}));

          await waitFor(() => {
            expect(mutation).toHaveBeenCalledWith(expectedValues);
          });
        });
      });
    });

    describe('with initial values', () => {
      const initialValues = {
        launchSite: 'KSC',
        mission: {
          name: 'Warlock IV',
        },
      };

      it('should submit the form', async () => {
        const mutation: Mutation = jest.fn();
        const mutationStatus: MutationStatus = { isLoading: false };
        const useMutation: UseMutation = () => [mutation, mutationStatus];

        const { getByRole } = render(
          <Form initialValues={initialValues} useMutation={useMutation}>
            <Fields />

            <button type="submit">Submit</button>
          </Form>
        );

        userEvent.click(getByRole('button', { name: 'Submit'}));

        await waitFor(() => {
          expect(mutation).toHaveBeenCalledWith(initialValues);
        });
      });

      describe('with params', () => {
        const params: QueryParams = { id: 0 };

        it('should submit the form', async () => {
          const mutation: Mutation = jest.fn();
          const mutationStatus: MutationStatus = { isLoading: false };
          const useMutation: UseMutation = () => [mutation, mutationStatus];

          const { getByRole } = render(
            <Form params={params} initialValues={initialValues} useMutation={useMutation}>
              <Fields />

              <button type="submit">Submit</button>
            </Form>
          );

          userEvent.click(getByRole('button', { name: 'Submit'}));

          await waitFor(() => {
            expect(mutation).toHaveBeenCalledWith({ ...initialValues, ...params });
          });
        });
      });
    });

    describe('with params', () => {
      const params: QueryParams = { id: 0 };

      it('should submit the form', async () => {
        const mutation: Mutation = jest.fn();
        const mutationStatus: MutationStatus = { isLoading: false };
        const useMutation: UseMutation = () => [mutation, mutationStatus];

        const { getByRole } = render(
          <Form initialValues={defaultValues} params={params} useMutation={useMutation}>
            <Fields />

            <button type="submit">Submit</button>
          </Form>
        );

        userEvent.click(getByRole('button', { name: 'Submit'}));

        await waitFor(() => {
          expect(mutation).toHaveBeenCalledWith({ ...defaultValues, ...params });
        });
      });
    });
  });
});
