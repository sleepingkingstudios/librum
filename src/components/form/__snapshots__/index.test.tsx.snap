// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="relative"
  >
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<Form /> when the form is loading should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="relative"
  >
    <div
      class="form-loading-overlay absolute grid h-full place-content-center w-full"
    >
      <div
        class="inline-block text-center"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-gear animate-spin motion-reduce:animate-none h-10 w-10"
          data-icon="gear"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M495.9 166.6C499.2 175.2 496.4 184.9 489.6 191.2L446.3 230.6C447.4 238.9 448 247.4 448 256C448 264.6 447.4 273.1 446.3 281.4L489.6 320.8C496.4 327.1 499.2 336.8 495.9 345.4C491.5 357.3 486.2 368.8 480.2 379.7L475.5 387.8C468.9 398.8 461.5 409.2 453.4 419.1C447.4 426.2 437.7 428.7 428.9 425.9L373.2 408.1C359.8 418.4 344.1 427 329.2 433.6L316.7 490.7C314.7 499.7 307.7 506.1 298.5 508.5C284.7 510.8 270.5 512 255.1 512C241.5 512 227.3 510.8 213.5 508.5C204.3 506.1 197.3 499.7 195.3 490.7L182.8 433.6C167 427 152.2 418.4 138.8 408.1L83.14 425.9C74.3 428.7 64.55 426.2 58.63 419.1C50.52 409.2 43.12 398.8 36.52 387.8L31.84 379.7C25.77 368.8 20.49 357.3 16.06 345.4C12.82 336.8 15.55 327.1 22.41 320.8L65.67 281.4C64.57 273.1 64 264.6 64 256C64 247.4 64.57 238.9 65.67 230.6L22.41 191.2C15.55 184.9 12.82 175.3 16.06 166.6C20.49 154.7 25.78 143.2 31.84 132.3L36.51 124.2C43.12 113.2 50.52 102.8 58.63 92.95C64.55 85.8 74.3 83.32 83.14 86.14L138.8 103.9C152.2 93.56 167 84.96 182.8 78.43L195.3 21.33C197.3 12.25 204.3 5.04 213.5 3.51C227.3 1.201 241.5 0 256 0C270.5 0 284.7 1.201 298.5 3.51C307.7 5.04 314.7 12.25 316.7 21.33L329.2 78.43C344.1 84.96 359.8 93.56 373.2 103.9L428.9 86.14C437.7 83.32 447.4 85.8 453.4 92.95C461.5 102.8 468.9 113.2 475.5 124.2L480.2 132.3C486.2 143.2 491.5 154.7 495.9 166.6V166.6zM256 336C300.2 336 336 300.2 336 255.1C336 211.8 300.2 175.1 256 175.1C211.8 175.1 176 211.8 176 255.1C176 300.2 211.8 336 256 336z"
            fill="currentColor"
          />
        </svg>
        <br />
        Loading...
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<Form /> with className: value should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="custom-form relative"
  >
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<Form /> with loadingAnimation: value when the form is loading should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="relative"
  >
    <div
      class="form-loading-overlay absolute grid h-full place-content-center w-full"
    >
      <div
        class="inline-block text-center"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-gear animate-pulse motion-reduce:animate-none h-10 w-10"
          data-icon="gear"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M495.9 166.6C499.2 175.2 496.4 184.9 489.6 191.2L446.3 230.6C447.4 238.9 448 247.4 448 256C448 264.6 447.4 273.1 446.3 281.4L489.6 320.8C496.4 327.1 499.2 336.8 495.9 345.4C491.5 357.3 486.2 368.8 480.2 379.7L475.5 387.8C468.9 398.8 461.5 409.2 453.4 419.1C447.4 426.2 437.7 428.7 428.9 425.9L373.2 408.1C359.8 418.4 344.1 427 329.2 433.6L316.7 490.7C314.7 499.7 307.7 506.1 298.5 508.5C284.7 510.8 270.5 512 255.1 512C241.5 512 227.3 510.8 213.5 508.5C204.3 506.1 197.3 499.7 195.3 490.7L182.8 433.6C167 427 152.2 418.4 138.8 408.1L83.14 425.9C74.3 428.7 64.55 426.2 58.63 419.1C50.52 409.2 43.12 398.8 36.52 387.8L31.84 379.7C25.77 368.8 20.49 357.3 16.06 345.4C12.82 336.8 15.55 327.1 22.41 320.8L65.67 281.4C64.57 273.1 64 264.6 64 256C64 247.4 64.57 238.9 65.67 230.6L22.41 191.2C15.55 184.9 12.82 175.3 16.06 166.6C20.49 154.7 25.78 143.2 31.84 132.3L36.51 124.2C43.12 113.2 50.52 102.8 58.63 92.95C64.55 85.8 74.3 83.32 83.14 86.14L138.8 103.9C152.2 93.56 167 84.96 182.8 78.43L195.3 21.33C197.3 12.25 204.3 5.04 213.5 3.51C227.3 1.201 241.5 0 256 0C270.5 0 284.7 1.201 298.5 3.51C307.7 5.04 314.7 12.25 316.7 21.33L329.2 78.43C344.1 84.96 359.8 93.56 373.2 103.9L428.9 86.14C437.7 83.32 447.4 85.8 453.4 92.95C461.5 102.8 468.9 113.2 475.5 124.2L480.2 132.3C486.2 143.2 491.5 154.7 495.9 166.6V166.6zM256 336C300.2 336 336 300.2 336 255.1C336 211.8 300.2 175.1 256 175.1C211.8 175.1 176 211.8 176 255.1C176 300.2 211.8 336 256 336z"
            fill="currentColor"
          />
        </svg>
        <br />
        Loading...
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<Form /> with loadingIcon: value when the form is loading should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="relative"
  >
    <div
      class="form-loading-overlay absolute grid h-full place-content-center w-full"
    >
      <div
        class="inline-block text-center"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-radiation h-10 w-10"
          data-icon="radiation"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 303.1c26.5 0 48-21.5 48-48S282.5 207.1 256 207.1S208 229.5 208 255.1S229.5 303.1 256 303.1zM213.6 188L142.7 74.71C132.5 58.41 109.9 54.31 95.25 66.75c-44.94 38.1-76.19 91.82-85.17 152.8C7.266 238.7 22.67 255.8 42.01 255.8h133.8C175.8 227.2 191 202.3 213.6 188zM416.8 66.75c-14.67-12.44-37.21-8.338-47.41 7.965L298.4 188c22.6 14.3 37.8 39.2 37.8 67.8h133.8c19.34 0 34.74-17.13 31.93-36.26C492.9 158.6 461.7 104.8 416.8 66.75zM298.4 323.5C286.1 331.2 271.6 335.9 256 335.9s-30.1-4.701-42.4-12.4L142.7 436.9c-10.14 16.21-4.16 38.2 13.32 45.95C186.6 496.4 220.4 504 256 504s69.42-7.611 100-21.18c17.48-7.752 23.46-29.74 13.32-45.95L298.4 323.5z"
            fill="currentColor"
          />
        </svg>
        <br />
        Loading...
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<Form /> with loadingMessage: value when the form is loading should match the snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="relative"
  >
    <div
      class="form-loading-overlay absolute grid h-full place-content-center w-full"
    >
      <div
        class="inline-block text-center"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-gear animate-spin motion-reduce:animate-none h-10 w-10"
          data-icon="gear"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M495.9 166.6C499.2 175.2 496.4 184.9 489.6 191.2L446.3 230.6C447.4 238.9 448 247.4 448 256C448 264.6 447.4 273.1 446.3 281.4L489.6 320.8C496.4 327.1 499.2 336.8 495.9 345.4C491.5 357.3 486.2 368.8 480.2 379.7L475.5 387.8C468.9 398.8 461.5 409.2 453.4 419.1C447.4 426.2 437.7 428.7 428.9 425.9L373.2 408.1C359.8 418.4 344.1 427 329.2 433.6L316.7 490.7C314.7 499.7 307.7 506.1 298.5 508.5C284.7 510.8 270.5 512 255.1 512C241.5 512 227.3 510.8 213.5 508.5C204.3 506.1 197.3 499.7 195.3 490.7L182.8 433.6C167 427 152.2 418.4 138.8 408.1L83.14 425.9C74.3 428.7 64.55 426.2 58.63 419.1C50.52 409.2 43.12 398.8 36.52 387.8L31.84 379.7C25.77 368.8 20.49 357.3 16.06 345.4C12.82 336.8 15.55 327.1 22.41 320.8L65.67 281.4C64.57 273.1 64 264.6 64 256C64 247.4 64.57 238.9 65.67 230.6L22.41 191.2C15.55 184.9 12.82 175.3 16.06 166.6C20.49 154.7 25.78 143.2 31.84 132.3L36.51 124.2C43.12 113.2 50.52 102.8 58.63 92.95C64.55 85.8 74.3 83.32 83.14 86.14L138.8 103.9C152.2 93.56 167 84.96 182.8 78.43L195.3 21.33C197.3 12.25 204.3 5.04 213.5 3.51C227.3 1.201 241.5 0 256 0C270.5 0 284.7 1.201 298.5 3.51C307.7 5.04 314.7 12.25 316.7 21.33L329.2 78.43C344.1 84.96 359.8 93.56 373.2 103.9L428.9 86.14C437.7 83.32 447.4 85.8 453.4 92.95C461.5 102.8 468.9 113.2 475.5 124.2L480.2 132.3C486.2 143.2 491.5 154.7 495.9 166.6V166.6zM256 336C300.2 336 336 300.2 336 255.1C336 211.8 300.2 175.1 256 175.1C211.8 175.1 176 211.8 176 255.1C176 300.2 211.8 336 256 336z"
            fill="currentColor"
          />
        </svg>
        <br />
        Activating Reactor...
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
